port/unix/omrmmap.c

```diff
@@ -50,11 +49,93 @@
 #include "protect_helpers.h"
 #include "omrportpriv.h"
 
+#if defined(J9ZOS390)
+#include "omrvmem.h"
+#if !defined(MAP_FAILED)
+#define MAP_FAILED ((void *)(intptr_t)-1)
+#endif /* !defined(MAP_FAILED) */
+#endif /* defined(J9ZOS390) */
+
 #if defined(AIXPPC)
 #include <sys/shm.h>
 #include <sys/vminfo.h>
 #endif/*AIXPPC*/
 
+#if defined(J9ZOS390)
+static J9MmapHandle *
+omrmmap_read_map_file(struct OMRPortLibrary *portLibrary, intptr_t file, uint64_t offset, uintptr_t size, uint32_t category)
+{
+	/* default implementation will allocate memory and read the file into it */
+	uintptr_t numBytesRead = 0;
+	intptr_t rc = 0;
+	void *mappedMemory = NULL;
+	void *allocPointer = NULL;
+	J9MmapHandle *returnVal = NULL;
+
+	Trc_PRT_mmap_map_file_default_entered(file, size);
+
+	if (-1 == file) {
+		Trc_PRT_mmap_map_file_default_badfile();
+		return NULL;
+	}
+
+	if (0 == size) {
+		struct stat buf;
+
+		memset(&buf, 0, sizeof(buf));
+		if (-1 == fstat(file - FD_BIAS, &buf)) {
+			Trc_PRT_mmap_map_file_unix_filestatfailed();
```

Comment:
TraceExit-Exception

Reply:
There are 6 places the function return NULL on error cases. Rather than changing all these `TraceException` into `TraceExit-Exception`, I will change them to use goto.

Reply:
Be aware that only the two return locations I commented on don't use `TraceExit` tracepoints; we don't want the reverse problem of too many exit tracepoints.

Reply:
> Be aware that only the two return locations I commented on don't use TraceExit tracepoints; we don't want the reverse problem of too many exit tracepoints.

I just realized this. Updated to use `TraceExit-Exception` trace point here.